### ISSUE TITLE
Incompatible types error with different versions for documents generated with code generators

### ISSUE THREAD

**Issue by xlight05** [ballerina-platform/ballerina-lang#36068]
**Description:**
Came across this issue when I was testing the azure functions package with latest update 1 pack. I published the bala to local repo. and added the entry to Ballerina.toml. when i built the project, I got the following error. it comes from the generated code from azure functions. 

```
ERROR [az-func-generated_1.bal:(3:46,3:51)] incompatible types: expected '(ballerina/http:2.3.0:Listener|error)', found 'ballerina/http:2.2.1:Listener'
ERROR [az-func-generated_1.bal:(4:55,4:62)] incompatible types: expected 'ballerina/http:2.2.1:Request', found 'ballerina/http:2.3.0:Request'
ERROR [az-func-generated_1.bal:(4:63,4:71)] incompatible types: expected 'ballerina/http:2.2.1:Response', found 'ballerina/http:2.3.0:Response'
ERROR [az-func-generated_1.bal:(8:55,8:62)] incompatible types: expected 'ballerina/http:2.2.1:Request', found 'ballerina/http:2.3.0:Request'
...
```

The generated code contains a http service. you can find a sample code below.
```
import ballerinax/azure_functions as af;
import ballerina/http;

public listener http:Listener __testListener = af:hl;

service http:Service / on __testListener {
    isolated resource function 'default hello(http:Caller caller, http:Request request) returns error? {
    }
}
```

Versions - 
http version in azure functions build.gradle - 2.2.1
http version in update 1 pack - 2.3.0

However when I added import ballerina/http as _;  as @azinneera suggested to the client code, issue went away.

**Steps to reproduce:**
Install update 1 pack.
Build azure functions modules with 2201.1.0-rc1.3 lang version
Publish to local
Build a sample azure functions code with the publishd bala in ballerina.toml

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

**Comment by azinneera** [comment]
Analyzed this and understood the following:
**_Use case:_**
1. User has imported `ballerinax/azure_functions` and `ballerinax/azure_functions` depends on `ballerina/http`
2. ballerinax/azure has a code generator compiler plugin and this plugin adds a new file that contains an import for `ballerina/http`

**_Compiler extension design:_**
According to the compiler extension design, compiler plugins are executed after the compilation and if there are any code generators or modifiers, the compilation will be run twice in the CLI, that is, before running the compiler plugins and after running the compiler plugins. Since the is a file edit operation taking place here, the compilation is done on a fresh package which runs the package resolution as well.

**_Root cause:_**
During the 1st compilation, `ballerina/http` is a transitve dependency so, dependency resolution adds the `http` version locked in the `ballerinax/azure_functions as af` (ballerina/http:2.2.1)
During the 2nd compilation, `ballerina/http` is a direct dependency and so it sticks to the latest version available (`ballerina/http:2.2.1`)
Since the second compilation is done on a new package instance, the compilation is expected to be a fresh compilation, but instead looks like some caches are not cleared from the first compilation.

Looping in @hasithaa and @MaryamZi

**Comment by sameerajayasoma** [comment]
We had a quick chat on this yesterday. 

Here is how the current dependency resolver works:
- Create the package dependency graph
- Store compiled representations of those dependencies in a package cache. 
- Reuse them in subsequent compilations (e.g., triggered via source code changes)
- Then, create a new dependency graph for each source code change. 
- Compare the new graph with the old graph to figure out package dependencies that need to be removed. 
- Update package cache by removing packages identified in the above step. 

To fix the above issue, my suggestion is to remove the unused packages and the packages that depend on them.
